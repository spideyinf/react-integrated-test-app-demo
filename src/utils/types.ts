export interface Movie {
  id: string;
  title: string;
  url: string;
  sharedBy: string;
  description: string;
}

export interface MovieContextData {
  movies: Movie[];
  user: any;
  appLoading: boolean;
  setMovies: (movies: Movie[]) => void;
}
