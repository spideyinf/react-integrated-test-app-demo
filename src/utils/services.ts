import { auth, db } from './firebase';
import { collection, getDocs, query, addDoc, orderBy, setDoc, doc } from '@firebase/firestore';
import { createUserWithEmailAndPassword, signInWithEmailAndPassword } from '@firebase/auth';
import { Movie } from './types';

const movies = collection(db, 'movies');

export const getYoutubeId = (url: string) => {
  let ID: string | string[] = '';
  const parsedUrl = url.replace(/(>|<)/gi, '').split(/(vi\/|v=|\/v\/|youtu\.be\/|\/embed\/)/);
  if (parsedUrl[2] !== undefined) {
    return parsedUrl[2].split(/[^0-9a-z_-]/i)[0];
  }
  return ID;
};

export const getMovies = async () => {
  try {
    const q = query(movies, orderBy('createdTime', 'desc'));
    const querySnapshot = await getDocs(q);
    const data: any = [];

    querySnapshot.forEach(async (doc) => {
      data.push({
        id: doc.id,
        ...doc.data()
      });
    });
    return data;
  } catch (error) {
    return [];
  }
};

export const shareMovie = (data: any) => {
  return addDoc(movies, {
    ...data,
    createdTime: new Date().toISOString()
  });
};

export const updateMovieService = (movie: Movie) => {
  const { id, ...data } = movie;
  return setDoc(doc(db, 'movies', id), data, { merge: true });
};

const register = async (email: string, password: string) => {
  try {
    await createUserWithEmailAndPassword(auth, email, password);
    await signInWithEmailAndPassword(auth, email, password);
  } catch (error) {
    alert('Login failed, check your credentials');
  }
};

export const loginRegister = async (email: string, password: string) => {
  try {
    await signInWithEmailAndPassword(auth, email, password);
  } catch (error: any) {
    const { message } = error;
    if (message && message.includes('auth/user-not-found')) {
      await register(email, password);
    } else {
      alert('Login failed, check your credentials');
    }
  }
};

export const fetchVideoInformation = async (url: string) => {
  try {
    const id = getYoutubeId(url);
    if (!id) return null;
    const data = await fetch(
      `https://www.googleapis.com/youtube/v3/videos?part=snippet&key=${process.env.REACT_APP_YOUTUBE_API_KEY}&id=${id}`
    );
    const result = await data.json();
    if (!result.items || !result.items.length) {
      return null;
    }
    return result;
  } catch (error) {
    console.log('error', error);
  }
};

export const logOut = async () => {
  try {
    await auth.signOut();
  } catch (error) {
    console.log(error);
  }
};
