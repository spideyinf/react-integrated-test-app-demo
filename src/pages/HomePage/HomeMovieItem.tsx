import { FC } from 'react';
import { Movie } from '../../utils/types';

interface Props {
  movie: Movie;
}

function classNames(...classes: string[]) {
  return classes.filter(Boolean).join(' ');
}

const HomeMovieItem: FC<Props> = ({ movie }) => {
  return (
    <li key={movie.id} className="col-span-1 flex shadow-sm rounded-md max-w-5xl">
      <iframe
        data-testid="movieIframe"
        className={classNames(
          '',
          'flex-shrink-0 flex items-center justify-center text-white text-sm font-medium rounded-l-md'
        )}
        src={movie.url}
        title={movie.title}
        height="240"
        width="360"
      />
      <div className="flex-1 flex items-center justify-between border-t border-r border-b border-gray-200 bg-white rounded-r-md truncate">
        <div className="px-4 py-2 truncate flex flex-col">
          <a href={movie.url} className="text-indigo-700 text-2xl font-bold hover:text-indigo-600">
            {movie.title}
          </a>
          <p className="text-gray-600">Shared by: {movie.sharedBy}</p>
          <p className="mt-2 text-md tex-gray-600">Description: </p>
          <p data-testid="description" className="text-gray-500 truncate">
            {movie.description}
          </p>
        </div>
      </div>
    </li>
  );
};

export default HomeMovieItem;
