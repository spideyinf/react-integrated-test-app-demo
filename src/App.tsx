import React, { createContext, useEffect, useState } from 'react';
import { Routes, Route } from 'react-router-dom';
import { Layout } from './components/Layout';
import HomePage from './pages/HomePage';
import NotFoundPage from './pages/NotFoundPage';
import SharePage from './pages/SharePage';
import { auth } from './utils/firebase';
import { Movie, MovieContextData } from './utils/types';

export const MoviesContext = createContext<MovieContextData>({
  movies: [],
  user: null,
  appLoading: true,
  setMovies: () => null
});

function App() {
  const [movies, setMovies] = useState<Movie[]>([]);
  const [appLoading, setAppLoading] = useState(true);
  const [user, setUser] = useState<any>(null);

  useEffect(() => {
    auth.onAuthStateChanged(async (authUser) => {
      if (authUser) {
        setUser(authUser);
      }
      setAppLoading(false);
    });
  }, []);

  return (
    <React.Suspense fallback={null}>
      <MoviesContext.Provider
        value={{
          movies,
          setMovies: (data: Movie[]) => setMovies(data),
          appLoading,
          user
        }}
      >
        <Layout>
          <Routes>
            <Route path="/" element={<HomePage />} />
            <Route path="/share" element={<SharePage />} />
            <Route path="/*" element={<NotFoundPage />} />
          </Routes>
        </Layout>
      </MoviesContext.Provider>
    </React.Suspense>
  );
}

export default App;
