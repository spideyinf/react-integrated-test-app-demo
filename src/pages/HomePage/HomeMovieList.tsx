import { FC, useContext } from 'react';
import { MoviesContext } from '../../App';
import HomeMovieItem from './HomeMovieItem';

export const HomeMovieList: FC = () => {
  const { movies } = useContext(MoviesContext);

  return (
    <ul className="flex flex-col items-center pt-8 space-y-8">
      {movies.length > 0 ? (
        movies.map((movie) => <HomeMovieItem key={movie.id} movie={movie} />)
      ) : (
        <h2 className="text-center pt-12 text-xl">No movies found</h2>
      )}
    </ul>
  );
};
