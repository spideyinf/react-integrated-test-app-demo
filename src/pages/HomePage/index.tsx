import { FC, useContext, useEffect, useState } from 'react';
import { MoviesContext } from '../../App';
import { getMovies } from '../../utils/services';
import { HomeMovieList } from './HomeMovieList';

const HomePage: FC = () => {
  const [loading, setLoading] = useState(false);
  const { movies, setMovies } = useContext(MoviesContext);

  useEffect(() => {
    if (!movies.length) setLoading(true);
    (async () => {
      const data = await getMovies();
      setMovies(data);
      setLoading(false);
    })();
  }, []);

  if (loading) return <h2 className="text-center pt-12 text-xl">Loading...</h2>;

  return <HomeMovieList />;
};

export default HomePage;
