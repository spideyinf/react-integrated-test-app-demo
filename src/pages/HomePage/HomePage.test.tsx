import { cleanup, render, screen, waitFor } from '@testing-library/react';
import { MoviesContext } from '../../App';
import { Movie } from '../../utils/types';
import HomeMovieItem from './HomeMovieItem';
import { HomeMovieList } from './HomeMovieList';
import HomePage from './HomePage';

jest.mock('@firebase/firestore', () => ({
  getDocs() {
    return Promise.resolve([
      {
        id: 1,
        data: () => ({
          title: 'Demo',
          url: 'https://www.youtube.com/embed/id',
          description: 'Description',
          sharedBy: 'test@gmail.com'
        })
      }
    ]);
  },
  collection: jest.fn(),
  getFirestore: jest.fn(),
  setDoc() {
    return Promise.resolve({
      status: '200'
    });
  }
}));

beforeEach(() => {
  cleanup();
});

describe('Test Home Page', () => {
  test('Movie list should render empty text', () => {
    render(<HomeMovieList />);
    expect(screen.getByText('No movies found')).toBeInTheDocument();
  });

  test('Video data', async () => {
    let movies: Movie[] = [];
    const setMovies: any = jest.fn((data) => (movies = data));

    render(
      <MoviesContext.Provider value={{ movies, setMovies, appLoading: false, user: null }}>
        <HomePage />
      </MoviesContext.Provider>
    );
    await waitFor(() => {
      expect(setMovies).toHaveBeenCalled();
    });
  });

  test('HomeMovieItem Test', () => {
    let movie: Movie = {
      id: '1',
      title: 'Demo',
      url: 'https://www.youtube.com/embed/id',
      description: 'Description',
      sharedBy: 'test@gmail.com'
    };

    render(<HomeMovieItem movie={movie} />);
    expect(screen.getByTestId('movieIframe')).toBeInTheDocument();
    expect(screen.getByText(/Demo/i)).toBeInTheDocument();
    expect(screen.getByTestId('description')).toHaveTextContent(movie.description);
  });

  test('HomeMovieList Test', () => {
    let movies: Movie[] = [
      {
        id: '1',
        title: 'Video 1',
        url: 'https://www.youtube.com/embed/id',
        description: 'Description',
        sharedBy: 'test@gmail.com'
      },
      {
        id: '2',
        title: 'Video 2',
        url: 'https://www.youtube.com/embed/id2',
        description: 'Description2',
        sharedBy: 'test2@gmail.com'
      }
    ];

    const { container } = render(
      <MoviesContext.Provider
        value={{ movies, setMovies: jest.fn(), appLoading: false, user: null }}
      >
        <HomeMovieList />
      </MoviesContext.Provider>
    );
    expect(screen.getByTestId('homeMovieList')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
