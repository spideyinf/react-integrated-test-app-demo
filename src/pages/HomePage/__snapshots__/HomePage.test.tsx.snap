// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Home Page HomeMovieList Test 1`] = `
<ul
  class="flex flex-col items-center pt-8 space-y-8"
  data-testid="homeMovieList"
>
  <li
    class="col-span-1 flex shadow-sm rounded-md max-w-5xl"
  >
    <iframe
      class="flex-shrink-0 flex items-center justify-center text-white text-sm font-medium rounded-l-md"
      data-testid="movieIframe"
      height="240"
      src="https://www.youtube.com/embed/id"
      title="Video 1"
      width="360"
    />
    <div
      class="flex-1 flex items-center justify-between border-t border-r border-b border-gray-200 bg-white rounded-r-md truncate"
    >
      <div
        class="px-4 py-2 truncate flex flex-col"
      >
        <a
          class="text-indigo-700 text-2xl font-bold hover:text-indigo-600"
          href="https://www.youtube.com/embed/id"
        >
          Video 1
        </a>
        <p
          class="text-gray-600"
        >
          Shared by: 
          test@gmail.com
        </p>
        <p
          class="mt-2 text-md tex-gray-600"
        >
          Description: 
        </p>
        <p
          class="text-gray-500 truncate"
          data-testid="description"
        >
          Description
        </p>
      </div>
    </div>
  </li>
  <li
    class="col-span-1 flex shadow-sm rounded-md max-w-5xl"
  >
    <iframe
      class="flex-shrink-0 flex items-center justify-center text-white text-sm font-medium rounded-l-md"
      data-testid="movieIframe"
      height="240"
      src="https://www.youtube.com/embed/id2"
      title="Video 2"
      width="360"
    />
    <div
      class="flex-1 flex items-center justify-between border-t border-r border-b border-gray-200 bg-white rounded-r-md truncate"
    >
      <div
        class="px-4 py-2 truncate flex flex-col"
      >
        <a
          class="text-indigo-700 text-2xl font-bold hover:text-indigo-600"
          href="https://www.youtube.com/embed/id2"
        >
          Video 2
        </a>
        <p
          class="text-gray-600"
        >
          Shared by: 
          test2@gmail.com
        </p>
        <p
          class="mt-2 text-md tex-gray-600"
        >
          Description: 
        </p>
        <p
          class="text-gray-500 truncate"
          data-testid="description"
        >
          Description2
        </p>
      </div>
    </div>
  </li>
</ul>
`;
